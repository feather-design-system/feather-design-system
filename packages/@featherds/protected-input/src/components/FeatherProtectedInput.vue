<script lang="ts">
import { FeatherInput, FeatherInputProps } from "@featherds/input";
import PasswordIcon from "./PasswordIcon.vue";
import { useLabelProperty } from "@featherds/composables/LabelProperty";
import { toRef, h, defineComponent, PropType, DefineComponent } from "vue";

const LABELS = {
  show: "Show password",
  hide: "Hide password",
};
export const props = {
  ...FeatherInputProps,
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default: () => {
      return LABELS;
    },
  },
} as const;
export const emits = {
  "update:modelValue": (v: string) => true,
};
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  props,
  emits,
  data() {
    return {
      showPassword: false,
    };
  },
  watch: {
    showPassword() {
      (this.$refs.input as HTMLInputElement).focus();
    },
  },
  setup(props) {
    return {
      //wrapping in computed labels to avoid hideLabel conflict from inputwrapper
      computedLabels: useLabelProperty<typeof LABELS>(
        toRef(props, "labels"),
        LABELS
      ),
    };
  },
  render() {
    const icon = h(PasswordIcon, {
      modelValue: this.showPassword,
      hide: this.computedLabels.hideLabel.value,
      show: this.computedLabels.showLabel.value,
      "onUpdate:modelValue": (val) => {
        this.showPassword = val;
      },
    });
    const attrs = { ...this.$attrs };
    delete attrs.clear; //dont allow clear in password.
    delete attrs.maxlength; //dont allow maxlength in password.
    return h(
      FeatherInput as unknown as DefineComponent,
      {
        ...attrs,
        ...this.$props,
        type: this.showPassword ? "input" : "password",
        ref: "input",
      },
      {
        post: () => icon,
      }
    );
  },
});
</script>
