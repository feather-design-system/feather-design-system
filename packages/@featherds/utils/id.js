const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
