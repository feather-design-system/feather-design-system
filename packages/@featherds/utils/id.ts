const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
