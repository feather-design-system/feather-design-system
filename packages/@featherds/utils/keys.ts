const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
