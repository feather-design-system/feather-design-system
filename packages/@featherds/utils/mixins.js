var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
