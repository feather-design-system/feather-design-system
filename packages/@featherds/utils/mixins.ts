const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
