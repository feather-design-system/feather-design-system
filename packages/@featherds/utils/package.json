{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
