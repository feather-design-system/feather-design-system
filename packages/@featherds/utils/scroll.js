var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
