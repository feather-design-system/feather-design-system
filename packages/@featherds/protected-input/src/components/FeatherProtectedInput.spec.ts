import FeatherProtectedInput from "./FeatherProtectedInput.vue";
import { mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import "@featherds/input-helper/test/MutationObserver";

jest.spyOn(id, "getSafeId").mockImplementation((x) => x);
const getWrapper = function (options: Record<string, unknown> = {}) {
  options.props = {
    label: "TEST",
  };
  return mount(FeatherProtectedInput, options);
};

describe("FeatherProtectedInput.vue", () => {
  it("should render with show password icon by default", () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should change input type when show password is clicked", async () => {
    const wrapper = getWrapper();
    await wrapper
      .find("[data-ref-id='feather-input-password-icon']")
      .trigger("click");
    expect(wrapper.element).toMatchSnapshot();
  });
});
