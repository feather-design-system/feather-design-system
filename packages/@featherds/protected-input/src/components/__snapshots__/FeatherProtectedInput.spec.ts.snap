// Vitest Snapshot v1

exports[`FeatherProtectedInput.vue > should change input type when show password is clicked 1`] = `
<div
  class="feather-input-container"
>
  <div
    class="feather-input-wrapper-container feather-input-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-input-label"
        >
          TEST
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <input
        aria-describedby="feather-input-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-input"
        data-ref-id="feather-input"
        id="feather-input-label"
        maxlength="false"
        name="feather-input-label"
        type="input"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
        <a
          class="password-icon hide-when-disabled"
          data-ref-id="feather-input-password-icon"
          href="#"
          title="Hide password"
        >
          <svg
            aria-hidden="true"
            class="feather-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.71,2.29A1,1,0,0,0,2.29,3.71l18,18a1,1,0,0,0,1.42,0,1,1,0,0,0,0-1.42Z"
            />
            <path
              d="M13.59,17.83A8.41,8.41,0,0,1,12,18a9.77,9.77,0,0,1-8.92-6A10.31,10.31,0,0,1,4.94,9.18L3.52,7.76a12.38,12.38,0,0,0-2.45,3.87,1,1,0,0,0,0,.74A11.79,11.79,0,0,0,12,20a11.5,11.5,0,0,0,3.27-.49Z"
            />
            <path
              d="M10.41,6.17A8.41,8.41,0,0,1,12,6a9.77,9.77,0,0,1,8.92,6,10.31,10.31,0,0,1-1.86,2.82l1.42,1.42a12.38,12.38,0,0,0,2.45-3.87,1,1,0,0,0,0-.74A11.79,11.79,0,0,0,12,4a11.5,11.5,0,0,0-3.27.49Z"
            />
          </svg>
        </a>
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherProtectedInput.vue > should render with show password icon by default 1`] = `
<div
  class="feather-input-container"
>
  <div
    class="feather-input-wrapper-container feather-input-content"
  >
    <div
      class="feather-input-border"
    >
      <div
        class="pre-border"
      />
      <div
        class="label-border"
        style="width: 0px;"
      >
        <label
          class="feather-input-label"
          data-ref-id="feather-form-element-label"
          for="feather-input-label"
        >
          TEST
        </label>
      </div>
      <div
        class="post-border"
      />
    </div>
    <div
      class="feather-input-wrapper"
    >
      <div
        class="prefix"
      >
        
        
      </div>
      
      <input
        aria-describedby="feather-input-description"
        aria-disabled="false"
        aria-invalid="false"
        class="feather-input"
        data-ref-id="feather-input"
        id="feather-input-label"
        maxlength="false"
        name="feather-input-label"
        type="password"
      />
      
      <div
        class="post"
      >
        <!--v-if-->
        <!--v-if-->
        
        
        <a
          class="password-icon hide-when-disabled"
          data-ref-id="feather-input-password-icon"
          href="#"
          title="Show password"
        >
          <svg
            aria-hidden="true"
            class="feather-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22.93,11.63A11.79,11.79,0,0,0,12,4,11.79,11.79,0,0,0,1.07,11.63a1,1,0,0,0,0,.74A11.79,11.79,0,0,0,12,20a11.79,11.79,0,0,0,10.93-7.63A1,1,0,0,0,22.93,11.63ZM12,18a9.77,9.77,0,0,1-8.92-6A9.77,9.77,0,0,1,12,6a9.77,9.77,0,0,1,8.92,6A9.77,9.77,0,0,1,12,18Z"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </a>
        
        
      </div>
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-input-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;
